### PR TITLE
Rpp interpolation

### PR DESCRIPTION
**Pull request by Aposhian**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #2850 |
| Primary OS tested on | Ubuntu ||
| Robotic platform tested on | Custom robot in Gazebo |

---

## Description of contribution in a few bullet points
Adds parameter `use_lookahead_point_interpolation` that interpolates between poses on path to ensure the lookahead point is exactly the desired lookahead distance away from the robot. This leads to smoother commanded velocities for sparse paths.

## Description of documentation updates required from your changes
Add parameter docs.

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 1]
```diff
@@ -372,6 +378,45 @@ void RegulatedPurePursuitController::rotateToHeading(
   angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
 }
 
+geometry_msgs::msg::Point circleSegmentIntersection(
```
Comment (SteveMacenski) [inline nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:381]: A diagram would be helpful here as to how this interpolation is happening, but I think I understand the thinking in the diagram I attached below if that's the same as yours. 

You have pt i and pt i-1 that you're trying to find the exact intersection point between on the circle of the lookahead radius (pt E). You set Pt i and Pt i-1 as 2 points on a line segment and use the math associated with circle-to-line intersection points to find Pt E.

The couple of assumptions you make is that the line segment roughly approximates the curve of the path. For small distances, that's a good assumption. For further ones in sparse paths, I do wonder if we shouldn't do something more. But if you find that this is good enough, that's fine enough for me. Its basically linear interpolation.

I think there's probably an easier way to do this with just linear interpolation on the line segment, but this works. 

image

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 2]
```diff
@@ -372,6 +378,45 @@ void RegulatedPurePursuitController::rotateToHeading(
...
+      return p;
+    };
+  auto [x_min, x_max] = std::minmax(x1, x2);
+  auto [y_min, y_max] = std::minmax(y1, y2);
+  auto in_segment = [&](geometry_msgs::msg::Point p) {
+      return x_min <= p.x && p.x <= x_max &&
+             y_min <= p.y && p.y <= y_max;
+    };
+  auto intersection1 = intersection(1.0);
+  if (in_segment(intersection1)) {
```
Comment (SteveMacenski) [inline nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:413]: This can just be `if (x_min <= p.x && p.x <= x_max && y_min <= p.y && p.y <= y_max)` adding the lambda doesn't really make this more clear if its only used in 1 location

**Review by SteveMacenski — APPROVED** [review]
LGTM, took me a hot minute and a diagram to get what you were going for, but I understand. Might be worth adding a few comments inline to make it more clear to other users about how this is being modeled. Particularly w.r.t. the `1.0` and `-1.0` solutions and how this is setup. Might even be worth adding a link to the comment I left with the diagram so that users looking at the code later have a geometric look at it

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:381]: Comment is mostly here for posterity

File: nav2_regulated_pure_pursuit_controller/README.md [hunk 3]
```diff
@@ -76,6 +76,7 @@ Note: The maximum allowed time to collision is thresholded by the lookahead poin
 | `rotate_to_heading_min_angle` | The difference in the path orientation and the starting robot orientation to trigger a rotate in place, if `use_rotate_to_heading` is enabled. | 
 | `max_angular_accel` | Maximum allowable angular acceleration while rotating to heading, if enabled | 
 | `max_robot_pose_search_dist` | Maximum integrated distance along the path to bound the search for the closest pose to the robot. This is set by default to the maximum costmap extent, so it shouldn't be set manually unless there are loops within the local costmap. | 
+| `use_lookahead_point_interpolation` | Enable interpolation between poses on the path for lookahead point selection. Helps sparse paths to avoid inducing discontinuous commanded velocities |
```
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/README.md:79]: This does add additional computation (that I haven't quantified) that maybe isn't necessary for dense paths. So maybe it should be enabled by default, but I think it seems like a good idea to leave it as an option for those trying to run the controller in a resource constrained envrionment.

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_regulated_pure_pursuit_controller/README.md:79]: Fair enough - default on is good though

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 4]
```diff
@@ -372,6 +378,45 @@ void RegulatedPurePursuitController::rotateToHeading(
...
+  double r)
+{
+  auto x1 = p1.pose.position.x;
+  auto x2 = p2.pose.position.x;
+  auto y1 = p1.pose.position.y;
+  auto y2 = p2.pose.position.y;
+
+  auto dx = x2 - x1;
+  auto dy = y2 - y1;
+  auto dr2 = std::pow(dx, 2) + std::pow(dy, 2);
```
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:393]: Do you have evidence to back that up? From this trivial example, it looks like it compiles to `dx*dx` anyway with `-O1` on gcc: https://godbolt.org/z/4fsarxWhM

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp [hunk 5]
```diff
@@ -372,6 +378,45 @@ void RegulatedPurePursuitController::rotateToHeading(
   angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
 }
 
+geometry_msgs::msg::Point circleSegmentIntersection(
+  const geometry_msgs::msg::PoseStamped & p1,
+  const geometry_msgs::msg::PoseStamped & p2,
+  double r)
+{
+  auto x1 = p1.pose.position.x;
```
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:386]: Sure. Can do.

- [ ] Use explicit types for primitive types

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:381]: Can do
- [ ] Make `circleSegmentIntersection` a member method

Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:381]: Yes, I can add more documentation and clarity to the code. I just wanted to get this up here before I invested into that :smile: . Yes, you are understanding how it works correctly. As far as whether linear interpolation is good enough vs something more complex, it is good enough to solve my use-case, and it is an incremental improvement for the controller, so I think it is good enough. Beyond linear interpolation I think starts to get into the realm of path planning, but I think we should hold off trying to implement it until it is determined that there is a need.

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp (hunk 2 above)
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:413]: Great point: I think that is an artifact of an earlier revision.

- [ ] Refactor `in_segment` lambda

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:393]: With optimization flags, sure but if you remove those  there is a difference. We can't actually compile binaries for apt install with certain flags due to the hardware-specific optimizations that would be made. The optimizations would be for the server blade's architecture, not the downstream user's systems. While our building in Nav2 does by default use some level of optimization, that's not representative of what binary users are  going to be experiencing. There are subtle differences so I'm cautious about assuming that you're going to see all of those optimizations performed

File: nav2_regulated_pure_pursuit_controller/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/README.md:79]: - [ ] Change `use_lookahead_point_interpolation` to default `true`

File: nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp:393]: Ok, sure I can change it to `dx * dx`, but I think if someone compiles with `-O0` then they deserve slower performance :wink: .

- [ ] Replace `std::pow` with multiplication

**Comment by Aposhian** [comment]
I've added some more unit tests to the interpolation method, and it looks like I caught some edge cases that I need to fix.

File: nav2_regulated_pure_pursuit_controller/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_regulated_pure_pursuit_controller/README.md:79]: Also maybe just `use_interpolation` would be good. I think with documentation of the new parameter, that is sufficient and short

**Comment by SteveMacenski** [comment]
I see you removed the negative solution space, is it true that its always the positive solution, even if the path does a U-turn and the "real" solution is behind the robot pose?

File: nav2_regulated_pure_pursuit_controller/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aposhian) [reply nav2_regulated_pure_pursuit_controller/README.md:79]: - [ ] Change `use_lookahead_point_interpolation` to `use_interpolation`

**Comment by SteveMacenski** [comment]
That min/max check you had before seemed rational and doesn't take any significant compute time. Even if you don't find the opposite solution, you could at least reject the update to use the raw lookahead point vs. returning a totally wrong-direction-ed point. 

Though, that geometry tool you showed is illustrative (you have a bunch of these cool web tools in your pocket!) I think you're probably correct that that solution is generally true. Maybe I'm being overly cautious / concerned if there's a bug later down the line creating a subtle issue. Maybe I'm just being paranoid :laughing:

**Comment by Aposhian** [comment]
I believe I have addressed all comments. All unit tests are now passing, and I have added an interactive jupyter notebook to the docs folder to illustrate the interpolation algorithm, as well as additional comments. I also have opened the corresponding docs PR to add the parameter `use_interpolation`.

**Comment by SteveMacenski** [comment]
Some reason CI jobs have been a little flaky to trigger, just toggled to get CI to run

**Review by SteveMacenski — APPROVED** [review]
As always, I appreciate that you update / add tests for your PRs! It helps quite alot to keep the quality level high. This looks much better (and simpler!) after a round :-) 

Waiting on CI
